### PR TITLE
Second pass at error rewrites (interpreter errors).

### LINKED ISSUE #3001: Reduce traceback size in user-facing errors
Using the following example:

```python
import numba as nb

def make_closure(x):
    @nb.njit
    def foo(y):
        x = x + 1  # bad code, not actually a closure
        return x + y
    return foo

foo = make_closure(1)
foo(2)
```

This generates the following traceback in Python 3:

```python-traceback
﻿Traceback (most recent call last):
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 114, in get
    return self._con[name]
KeyError: 'x'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 719, in get_exact
    return self.localvars.get(name)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 116, in get
    raise NotDefinedError(name)
numba.errors.NotDefinedError: 'x' is not defined in None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 114, in get
    return self._con[name]
KeyError: 'x'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test2.py", line 12, in <module>
    foo(2)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/dispatcher.py", line 360, in _compile_for_args
    raise e
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/dispatcher.py", line 311, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/dispatcher.py", line 618, in compile
    cres = self._compiler.compile(args, return_type)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/dispatcher.py", line 83, in compile
    pipeline_class=self.pipeline_class)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 871, in compile_extra
    return pipeline.compile_extra(func)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 365, in compile_extra
    return self._compile_bytecode()
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 802, in _compile_bytecode
    return self._compile_core()
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 789, in _compile_core
    res = pm.run(self.status)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 251, in run
    raise patched_exception
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 243, in run
    stage()
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 379, in stage_analyze_bytecode
    func_ir = translate_stage(self.func_id, self.bc)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/compiler.py", line 935, in translate_stage
    return interp.interpret(bytecode)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/interpreter.py", line 110, in interpret
    self._dispatch(inst, kws)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/interpreter.py", line 239, in _dispatch
    raise e
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/interpreter.py", line 235, in _dispatch
    return fn(inst, **kws)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/interpreter.py", line 540, in op_LOAD_FAST
    self.store(value=self.get(srcname), name=res)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/interpreter.py", line 273, in get
    var = self.current_scope.get(name)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 711, in get
    return self.get_exact(name)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 722, in get_exact
    return self.parent.get(name)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 711, in get
    return self.get_exact(name)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 719, in get_exact
    return self.localvars.get(name)
  File "/Users/jcrist/anaconda/envs/skein/lib/python3.6/site-packages/numba/ir.py", line 116, in get
    raise NotDefinedError(name)
numba.errors.NotDefinedError: 'x' is not defined in test2.py (7)
```

The source of this looooooong trackeback is [exception chaining](https://www.python.org/dev/peps/pep-3134/) in Python 3, which includes previous exceptions when exceptions are raised in `except` blocks. While useful from a numba development point, the long tracebacks are difficult as a user. This is especially apparent in notebooks, where users often don't scroll down to the bottom of the exception, and instead see the earlier exceptions (`KeyError`, etc...).

A few potential solutions:

- Throw "clean" final exceptions using [`raise_from`](https://pythonhosted.org/six/#six.raise_from), which can trim chained exceptions. In this case, the `NotDefinedError` would be thrown as:

```python
raise_from(NotDefinedError(....), None)
```

- User facing exceptions are caught and re-raised in the user facing function (e.g. `njit`), significantly reducing the size of the traceback. This might look like:

```python
try:
    return do_numba_stuff(...)
except ExceptionsThatShouldBeTrimmedBaseClass as exc:
    raise_from(exc, None)
```

In both cases the clean re-raising could also be turned off in "debug mode", allowing for better UX without inhibiting development.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This is a second pass at improving error handling, the first pass
focussed on Typing and related errors, this pass focusses on
errors arising during bytecode interpretation and analysis.

Fixes #3001.

### CODE REVIEW

File: numba/errors.py [hunk 1]
```diff
@@ -241,6 +241,35 @@ def termcolor():
 https://github.com/numba/numba/issues/new
 """
 
+interpreter_error_info = """
+Unsupported Python functionality was found in the code Numba was trying to
+compile. This error could be due to invalid code, does the code work
+without Numba? (A quick way to check is to temporarily disable Numba JIT by
```
Comment (seibert) [inline numba/errors.py]: Can also shorten this to "(To temporarily disable the Numba JIT, set ... and then rerun the code.)"

File: numba/errors.py [hunk 2]
```diff
@@ -241,6 +241,35 @@ def termcolor():
...
+"""
+
+constant_inference_info = """
+Numba could not make a constant out of something that it decided should be
+a constant. This could well be a current limitation in Numba's internals,
+please either raise a bug report along with a minimal reproducer at:
+https://github.com/numba/numba/issues/new
+
+(if you need help writing a minimal reproducer please see)
+http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports
```
Comment (seibert) [inline numba/errors.py]: The note about minimal reproducers is redundant to instructions in the issue tracker, so I think we can drop these 2 lines from the error message.

File: numba/errors.py [hunk 3]
```diff
@@ -241,6 +241,35 @@ def termcolor():
 https://github.com/numba/numba/issues/new
 """
 
+interpreter_error_info = """
+Unsupported Python functionality was found in the code Numba was trying to
+compile. This error could be due to invalid code, does the code work
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/errors.py]: hmmm, it could also be due to the IR parse/build failing. This is unlikely but may happen?

File: numba/errors.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/errors.py]: Agreed. I've removed this from here and the other "help" strings.

File: numba/errors.py [hunk 4]
```diff
@@ -241,6 +241,35 @@ def termcolor():
...
+a constant. This could well be a current limitation in Numba's internals,
+please either raise a bug report along with a minimal reproducer at:
+https://github.com/numba/numba/issues/new
+
+(if you need help writing a minimal reproducer please see)
+http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports
+
+Or, alternatively please feel free to speak to the Numba core developers
+directly at: https://gitter.im/numba/numba
+"""
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/errors.py]: Agreed. I am only leaving this in for the case of a "reportable error" help string on the basis that these errors are possibly more problematic and some engagement with core devs might be needed.

File: numba/errors.py [hunk 5]
```diff
@@ -399,48 +430,58 @@ class UnsupportedError(NumbaError):
...
-        loc = "?" if self.loc is None else self.loc
-        return "{name!r} is not defined in {loc}".format(name=self.name,
-                                                         loc=self.loc)
+        msg = "Variable '%s' is not defined." % name
+        super(NotDefinedError, self).__init__(msg, loc=loc)
 
 
 class VerificationError(IRError):
+    """
+    An error occurred during IR verification.
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/errors.py]: yes. patched.
